packages/docusaurus-theme-classic/src/theme/ThemedImage/index.tsx

```diff
@@ -0,0 +1,23 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React from 'react';
+import type {Props} from '@theme/ThemedImage';
+import useThemeContext from '@theme/hooks/useThemeContext';
+
+function ThemedImage(props: Props): JSX.Element {
+  const {isDarkTheme} = useThemeContext();
+  const {darkSrc, lightSrc, alt = '', ...propsRest} = props;
+
+  if (darkSrc && isDarkTheme) {
+    return <img src={darkSrc} alt={alt} {...propsRest} />;
+  } else {
+    return <img src={lightSrc} alt={alt} {...propsRest} />;
+  }
+}
+
+export default ThemedImage;
```

Comment:
This component will only render one image (the light one) on the other on server-rendered markup.

See `useTheme` hook:

```tsx
const getInitialTheme = () => {
  if (!ExecutionEnvironment.canUseDOM) {
    return themes.light; // SSR: we don't care
  }
  return coerceToTheme(document.documentElement.getAttribute('data-theme'));
};
```

On hydration, the light image will be replaced with the dark one, producing an unwanted flash.

The only viable solution to fix the problem is to always render the 2 images, and display one or the other thanks to CSS (which is loaded before the user actually see anything, so there's no flash)

```css
img.themed-image{
  display: none;
}

html[data-theme='light'] img.themed-image.themed-image--light {
  display: block;
}
html[data-theme='dark'] img.themed-image.themed-image--dark {
  display: block;
}
```

Note: it's worth considering that light/dark is common nowadays but the CSS spec says this might be expanded to more values in the future. Having an API surface that is more extensible could be more future-proof (ie user can declare its own theme name, we should rather have `sources={dark: darkSrc,light:lightSrc,sepia: sepiaSrc}`)